// @gluestack-style/react/lib/module/resolver/checkAndPush.js
import { deepMergeArray } from '../utils';
export function checkAndPush(item, ret, keyToCheck) {
  function getIndexes(array, str) {
    return array.map((item, index) => item === str ? index : -1).filter(i => i !== -1);
  }
  function createNestedObject(arr) {
    let obj = {};
    arr.reduce((acc, curr) => {
      return acc[curr] = {};
    }, obj);
    return obj;
  }
  function setNestedObjectValue(obj, keyPath, value) {
    // If the key path is empty, return the value
    if (keyPath.length === 0) return value;

    // Otherwise, set the value at the current key path and recurse
    const key = keyPath[0];
    obj[key] = obj[key] || {};
    obj[key] = setNestedObjectValue(obj[key], keyPath.slice(1), value);
    return obj;
  }
  // keyToCheck = "baseStyle" | "variants" | "sizes"
  if (item.meta.path.includes(keyToCheck)) {
    // if (!item.meta.path.includes('state')) {
    //   if (!ret.ids) {
    //     ret.ids = [];
    //   }
    //   ret.ids.push(item.meta.cssId);
    //   ret.props = item?.meta?.props;
    // } else
    if (!item.meta.path.includes('state') && !item.meta.path.includes('colorMode')) {
      var _item$meta;
      if (!ret.ids) {
        ret.ids = [];
      }
      ret.ids.push(item.meta.cssId);
      ret.props = item === null || item === void 0 || (_item$meta = item.meta) === null || _item$meta === void 0 ? void 0 : _item$meta.props;

      // ret.default.push(item.meta.cssId);
    } else if (item.meta.path.includes('state') || item.meta.path.includes('colorMode')) {
      var _item$meta2;
      const allStates = getIndexes(item.meta.path, 'state');
      const allColorModes = getIndexes(item.meta.path, 'colorMode');

      // const allStatesAndColorMode = [...allStates, ...allColorModes];

      const mergeAllStateKey = [];
      allStates.forEach(statePath => {
        const state = item.meta.path[statePath + 1];
        mergeAllStateKey.push('state');
        mergeAllStateKey.push(state);
      });
      allColorModes.forEach(colorModePath => {
        const colorMode = item.meta.path[colorModePath + 1];
        mergeAllStateKey.push('colorMode');
        mergeAllStateKey.push(colorMode);
      });
      const stateObject = createNestedObject(mergeAllStateKey);
      setNestedObjectValue(stateObject, mergeAllStateKey, {
        ids: [item.meta.cssId],
        props: item === null || item === void 0 || (_item$meta2 = item.meta) === null || _item$meta2 === void 0 ? void 0 : _item$meta2.props
      });
      deepMergeArray(ret, stateObject);
    }
  }
}
