// @gluestack-ui/slider/lib/module/SliderFilledTrack.js
function _extends() { _extends = Object.assign ? Object.assign.bind() : function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }
import React, { forwardRef } from 'react';
import { SliderContext } from './Context';
import { Platform } from 'react-native';
import { mergeRefs } from '@gluestack-ui/utils';
function SliderFilledTrack(StyledSliderFilledTrack) {
  return /*#__PURE__*/forwardRef((_ref2, ref) => {
    let {
      style,
      ...props
    } = _ref2;
    const _ref = React.useRef(null);
    const {
      state,
      trackLayout,
      orientation,
      isDisabled,
      isFocused,
      isHovered,
      isPressed,
      isFocusVisible,
      isHoveredProp,
      isDisabledProp,
      isFocusedProp,
      isFocusVisibleProp,
      isPressedProp
    } = React.useContext(SliderContext);
    const getSliderTrackPosition = () => {
      if (orientation === 'vertical') {
        return trackLayout.height * state.getThumbPercent(0);
      } else {
        return trackLayout.width * state.getThumbPercent(0);
      }
    };
    const sliderTrackPosition = getSliderTrackPosition();
    const positionProps = orientation === 'vertical' ? {
      height: sliderTrackPosition
    } : {
      width: sliderTrackPosition
    };
    return /*#__PURE__*/React.createElement(StyledSliderFilledTrack, _extends({}, props, {
      ref: mergeRefs([_ref, ref]),
      style: {
        ...style,
        ...positionProps
      },
      states: {
        hover: isHovered || isHoveredProp,
        disabled: isDisabled || isDisabledProp,
        focus: isFocused || isFocusedProp,
        focusVisible: isFocusVisible || isFocusVisibleProp,
        active: isPressed || isPressedProp
      },
      disabled: isDisabled,
      tabIndex: Platform.OS === 'web' ? -1 : undefined
    }));
  });
}
export default SliderFilledTrack;
