// @gluestack-style/react/lib/commonjs/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  styled: true,
  verboseStyled: true,
  resolveBuildTimeSx: true,
  StyledProvider: true,
  useStyled: true,
  Theme: true,
  useTheme: true,
  styledToStyledResolved: true,
  styledResolvedToOrderedSXResolved: true,
  flush: true,
  convertStyledToStyledVerbosed: true,
  convertSxToSxVerbosed: true,
  createStyled: true,
  createGlobalStylesWeb: true,
  propertyTokenMap: true,
  AsForwarder: true,
  createStyle: true,
  INTERNAL_updateCSSStyleInOrderedResolved: true,
  createConfig: true,
  createComponents: true
};
Object.defineProperty(exports, "AsForwarder", {
  enumerable: true,
  get: function () {
    return _AsForwarder.AsForwarder;
  }
});
Object.defineProperty(exports, "INTERNAL_updateCSSStyleInOrderedResolved", {
  enumerable: true,
  get: function () {
    return _updateCSSStyleInOrderedResolved.INTERNAL_updateCSSStyleInOrderedResolved;
  }
});
Object.defineProperty(exports, "StyledProvider", {
  enumerable: true,
  get: function () {
    return _StyledProvider.StyledProvider;
  }
});
Object.defineProperty(exports, "Theme", {
  enumerable: true,
  get: function () {
    return _Theme.Theme;
  }
});
Object.defineProperty(exports, "convertStyledToStyledVerbosed", {
  enumerable: true,
  get: function () {
    return _convertSxToSxVerbosed.convertStyledToStyledVerbosed;
  }
});
Object.defineProperty(exports, "convertSxToSxVerbosed", {
  enumerable: true,
  get: function () {
    return _convertSxToSxVerbosed.convertSxToSxVerbosed;
  }
});
Object.defineProperty(exports, "createComponents", {
  enumerable: true,
  get: function () {
    return _createConfig.createComponents;
  }
});
Object.defineProperty(exports, "createConfig", {
  enumerable: true,
  get: function () {
    return _createConfig.createConfig;
  }
});
Object.defineProperty(exports, "createGlobalStylesWeb", {
  enumerable: true,
  get: function () {
    return _createGlobalStylesWeb.createGlobalStylesWeb;
  }
});
Object.defineProperty(exports, "createStyle", {
  enumerable: true,
  get: function () {
    return _createStyle.createStyle;
  }
});
Object.defineProperty(exports, "createStyled", {
  enumerable: true,
  get: function () {
    return _createStyled.createStyled;
  }
});
Object.defineProperty(exports, "flush", {
  enumerable: true,
  get: function () {
    return _injectInStyle.flush;
  }
});
Object.defineProperty(exports, "propertyTokenMap", {
  enumerable: true,
  get: function () {
    return _propertyTokenMap.propertyTokenMap;
  }
});
Object.defineProperty(exports, "resolveBuildTimeSx", {
  enumerable: true,
  get: function () {
    return _styled.resolveBuildTimeSx;
  }
});
Object.defineProperty(exports, "styled", {
  enumerable: true,
  get: function () {
    return _styled.styled;
  }
});
Object.defineProperty(exports, "styledResolvedToOrderedSXResolved", {
  enumerable: true,
  get: function () {
    return _orderedResolved.styledResolvedToOrderedSXResolved;
  }
});
Object.defineProperty(exports, "styledToStyledResolved", {
  enumerable: true,
  get: function () {
    return _styledResolved.styledToStyledResolved;
  }
});
Object.defineProperty(exports, "useStyled", {
  enumerable: true,
  get: function () {
    return _StyledProvider.useStyled;
  }
});
Object.defineProperty(exports, "useTheme", {
  enumerable: true,
  get: function () {
    return _Theme.useTheme;
  }
});
Object.defineProperty(exports, "verboseStyled", {
  enumerable: true,
  get: function () {
    return _styled.verboseStyled;
  }
});
var _styled = require("./styled");
var _StyledProvider = require("./StyledProvider");
var _Theme = require("./Theme");
var _styledResolved = require("./resolver/styledResolved");
var _orderedResolved = require("./resolver/orderedResolved");
var _injectInStyle = require("./injectInStyle");
var _convertSxToSxVerbosed = require("./convertSxToSxVerbosed");
var _createStyled = require("./createStyled");
var _createGlobalStylesWeb = require("./createGlobalStylesWeb");
var _propertyTokenMap = require("./propertyTokenMap");
var _AsForwarder = require("./AsForwarder");
var _plugins = require("./plugins");
Object.keys(_plugins).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _plugins[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _plugins[key];
    }
  });
});
var _createStyle = require("./createStyle");
var _updateCSSStyleInOrderedResolved = require("./updateCSSStyleInOrderedResolved");
var _createConfig = require("./createConfig");
var _core = require("./core");
Object.keys(_core).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _core[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _core[key];
    }
  });
});
var _hooks = require("./hooks");
Object.keys(_hooks).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _hooks[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _hooks[key];
    }
  });
});
